Formula/code-server.rb

```diff
@@ -0,0 +1,78 @@
+class CodeServer < Formula
+  desc "Access VS Code through the browser"
+  homepage "https://coder.com"
+  # We download a built linux release but rebuild the native parts against our version of node/darwin.
+  # This just makes things easier as we don't have to rebuild all of the javascript.
+  url "https://github.com/cdr/code-server/releases/download/3.1.1/code-server-3.1.1-linux-x86_64.tar.gz"
+  version "3.1.1"
+  sha256 "5dd922d28b2e351c146081849d987fb1e439ee7d53b941434b2eecb2a194da71"
+
+  depends_on "node@12"
+
+  def install
+    # Rebuild code-server native modules.
+    system "npm", "rebuild"
+
+    # Rebuild vscode native modules.
+    Dir.chdir("lib/vscode")
+    system "npm", "rebuild"
+    Dir.chdir("../..")
+
+    # We remove the bundled version of node as we want code-server to use our version which is node@12.
+    (buildpath/"node").delete
+
+    libexec.install Dir["*"]
+
+    # Use our version of node.
+    (bin/"code-server").write <<~EOS
+      #!/bin/bash
+      exec #{HOMEBREW_PREFIX}/opt/node@12/bin/node #{libexec}/out/node/entry.js "$@"
+    EOS
+
+    # Later on when the code-server startup script works for macOS, we can use this.
+    # And remove the script above.
+    # See https://github.com/cdr/code-server/issues/1537
+    # (libexec/"node").make_relative_symlink (HOMEBREW_PREFIX/"opt/node@12/bin/node")
+    # (bin/"code-server").make_relative_symlink (libexec/"code-server")
+  end
+
+  def caveats
+    <<~EOS
+      Logs, including the autogenerated password for the launchd service are located at #{var}/log/code-server.log
+    EOS
+  end
+
+  plist_options :manual => "code-server"
+
+  def plist
+    <<~EOS
+      <?xml version="1.0" encoding="UTF-8"?>
+      <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+      <plist version="1.0">
+      <dict>
+        <key>KeepAlive</key>
+        <true/>
+        <key>Label</key>
+        <string>#{plist_name}</string>
+        <key>ProgramArguments</key>
+        <array>
+          <string>#{bin}/code-server</string>
+        </array>
+        <key>RunAtLoad</key>
+        <true/>
+        <key>WorkingDirectory</key>
+        <string>#{ENV["HOME"]}</string>
+        <key>StandardOutPath</key>
+        <string>#{var}/log/code-server.log</string>
+        <key>StandardErrorPath</key>
+        <string>#{var}/log/code-server.log</string>
+      </dict>
+      </plist>
+    EOS
+  end
+
+  test do
+    # This will fail if the native modules were not recompiled.
+    system bin/"code-server", "--list-extensions"
```

Comment:
We need a test that actually tests the software as a user would.

Reply:
To be clear, you want at test that starts the server and hits an endpoint to make sure it works?

This is the test we use for all releases at code-server because it ensures the native modules work and code-server can run and list installed extensions.

Reply:
Yes, I want to make sure it can do a thing a user would do with it.